File: 0957-minimum-add-to-make-parentheses-valid/0957-minimum-add-to-make-parentheses-valid.cpp
class Solution {
public:
    int minAddToMakeValid(string s) {
        int open = 0, close = 0;
        
        for (char ch : s) {
            if (ch == '(') {
                // Increment for an opening parenthesis
                open++;
            } else {
                // If there's an unmatched opening parenthesis, close it
                if (open > 0) {
                    open--;
                } else {
                    // Else we need a closing parenthesis to balance
                    close++;
                }
            }
        }
        
        // The result is the sum of unmatched open and close parentheses
        return open + close;
    }
};

File: 0957-minimum-add-to-make-parentheses-valid/0957-minimum-add-to-make-parentheses-valid.py
class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        open_count = 0
        close_count = 0
        
        for ch in s:
            if ch == '(':
                # Increment for an opening parenthesis
                open_count += 1
            else:
                # If there's an unmatched opening parenthesis, close it
                if open_count > 0:
                    open_count -= 1
                else:
                    # Else we need a closing parenthesis to balance
                    close_count += 1

        # The result is the sum of unmatched open and close parentheses
        return open_count + close_count
